service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java

```diff
@@ -0,0 +1,296 @@
+    } catch (IOException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  private static Instance createInstanceModel(
+      FlightContext flightContext,
+      String instanceId,
+      String projectId,
+      String serviceAccountEmail,
+      String workspaceUserFacingId,
+      String cliServer,
+      String gitHash) {
+    Instance instance = new Instance();
+    ApiGcpGceInstanceCreationParameters creationParameters =
+        flightContext
+            .getInputParameters()
+            .get(CREATE_GCE_INSTANCE_PARAMETERS, ApiGcpGceInstanceCreationParameters.class);
+    setFields(
+        creationParameters,
+        instanceId,
+        serviceAccountEmail,
+        workspaceUserFacingId,
+        cliServer,
+        instance,
+        gitHash);
+    setNetworks(instance, projectId, flightContext.getWorkingMap());
+    return instance;
+  }
+
+  @VisibleForTesting
+  static Instance setFields(
+      ApiGcpGceInstanceCreationParameters creationParameters,
+      String instanceId,
+      String serviceAccountEmail,
+      String workspaceUserFacingId,
+      String cliServer,
+      Instance instance,
+      String gitHash) {
+    instance.setName(instanceId).setMachineType(creationParameters.getMachineType());
+
+    instance.setDisks(
+        List.of(
+            new AttachedDisk()
+                .setBoot(true)
+                .setAutoDelete(true)
+                .setInitializeParams(
+                    new AttachedDiskInitializeParams()
+                        .setSourceImage(creationParameters.getVmImage())
+                        .setDiskSizeGb(Long.valueOf(100)))));
+
+    instance.setServiceAccounts(
+        List.of(
+            new ServiceAccount().setEmail(serviceAccountEmail).setScopes(SERVICE_ACCOUNT_SCOPES)));
+
+    List<ApiGcpGceInstanceGuestAccelerator> guestAccelerators =
+        creationParameters.getGuestAccelerators();
+    if (guestAccelerators != null) {
+      instance.setGuestAccelerators(
+          guestAccelerators.stream()
+              .map(
+                  accelerator ->
+                      new AcceleratorConfig()
+                          .setAcceleratorType(accelerator.getType())
+                          .setAcceleratorCount(accelerator.getCardCount()))
+              .collect(Collectors.toList()));
+    }
+    // Set metadata
+    Map<String, String> metadata = new HashMap<>();
+    Optional.ofNullable(creationParameters.getMetadata()).ifPresent(metadata::putAll);
+    addDefaultMetadata(metadata, workspaceUserFacingId, cliServer);
+    instance.setMetadata(
+        new Metadata()
+            .setItems(
+                metadata.entrySet().stream()
+                    .map(i -> new Items().setKey(i.getKey()).setValue(i.getValue()))
+                    .collect(Collectors.toList())));
+    return instance;
+  }
+
+  private static void addDefaultMetadata(
```

Comment:
share this.

Reply:
The notebooks also have some proxy specific metadata and it's unclear to me if the GCP instances will need some sort of equivalent yet